Configuration actions

vojtechtrefny opened the pull request:
This is not a final version, just a "request for comments".

I've decided to use the "single class" approach with a dict of supported ("configurable") attributes and functions used to set the attribute. I didn't have to create a special function for label configuration -- we already have the `write_label` functions so I've just changed it to take the `dry_run` option for the input validation.

Pros:
- It's easy to tell if format supports configuration of an attribute using actions.
- Having only one class makes the API simpler for the users.
- It's possible to use existing validation functions/tasks.

Cons:
- Every attribute needs a special function that looks like a setter but it actually isn't (because of the `dry_run` option).

vpodzime commented:
Looks good to me.

dwlehman commented:
I notice this doesn't really do any explicit checking of the new value. I think any errors would come up in `apply`, which I'm not sure is what we want -- I think it might be better to check it from the constructor.

Also, I'm not sure what the value is of setting `self._run` in the action class.

Otherwise I think it looks good.

vojtechtrefny commented:
> I notice this doesn't really do any explicit checking of the new value. I think any errors would come up in apply, which I'm not sure is what we want -- I think it might be better to check it from the constructor.

I wans't sure if the checks should go to apply or to constructor, I'll move it.

> Also, I'm not sure what the value is of setting self._run in the action class.

I just didn't want to have the same code (getattr) twice in `apply` and `execute`. Maybe the name is a little misleading.

dwlehman commented:
Yes, I think if you named it `self._execute` or `self._commit` or something else fairly explicit it would be okay. I think you should also do something along the lines of `assert(callable(self._execute))` right after setting it.

dwlehman commented:
I also think we can do better than `conf_actions`. I'm not sure what would be best, but `config_actions` or maybe even `config_actions_map` or `config_action_methods` come to mind.

vojtechtrefny commented:
New version -- added ActionCofigureDevice, checks moved to action constructor, changed same attribute/function names and added some basic test for configure action functionality.

vpodzime commented on blivet/formats/fs.py:
```diff
@@ -77,7 +77,8 @@ class FS(DeviceFormat):
 
-    config_actions_map = {"label" : "write_label"}
+    config_actions_map = {"label": "write_label",
+                          "mountpoint": None}
```
What is the purpose and result of this action?

vojtechtrefny replied:
This is intended to change (future) mountpoint for existing devices during installation or some "installation mode". It indeed is quite a weird action because it does nothing during `execute`, it just set the attribute.

vpodzime commented:
Looks good to me other than the above question.

dwlehman commented on blivet/deviceaction.py:
```diff
@@ -981,3 +987,105 @@ def obsoletes(self, action):
+    def execute(self, callbacks=None):
+        super().execute(callbacks=callbacks)
+
+        if self._execute:
```
`if self._execute is not None:` would be better.

vojtechtrefny replied:
Fixed.

dwlehman commented:
Overall I like it, but I have some concern about the device and format configuration action classes being almost identical. I realize that trying to combine them into a single class brings its own issues (and complexity), but it seems like asking for trouble requiring every change to made in two classes rather than one.

vojtechtrefny commented:
I think it makes sense to have two actions because we already have two actions for everything -- API users probably might expect to have a different actions for devices and formats.

Maybe an abstract class with common functionality could be useful.

vojtechtrefny commented:
Removed the mountpoint configuration commit.

vpodzime commented:
Looks good to me.

vojtechtrefny commented:
One PEP8 fix for a different change was part of the removed commit, fixed.
